common/cache.go

```diff
@@ -13,90 +18,300 @@ import (
+	b := c.readBlockFromDisk(height)
+	if b == nil {
+		// Some kind of disk corruption occured, start over
+		c.truncZero()
+	}
+	return b
+}
 
-	mutex sync.RWMutex
+func (c *BlockCache) setLatestHash() {
+	c.latestHash = nil
+	// There is at least one block; get the last block's hash
+	if c.nextBlock > c.firstBlock {
+		// At least one block remains; get the last block's hash
+		block := c.readBlock(c.nextBlock - 1)
+		if block == nil {
+			return
+		}
+		c.latestHash = make([]byte, len(block.Hash))
+		copy(c.latestHash, block.Hash)
+	}
 }
 
 // NewBlockCache returns an instance of a block cache object.
-func NewBlockCache(maxEntries int) *BlockCache {
-	return &BlockCache{
-		MaxEntries: maxEntries,
-		m:          make(map[int]*blockCacheEntry),
+func NewBlockCache(dbPath string, chainName string, startHeight int, redownload bool, backup int) *BlockCache {
+	c := &BlockCache{}
+	c.firstBlock = startHeight
+	c.nextBlock = startHeight
+	c.lengthsName, c.blocksName = dbFileNames(dbPath, chainName)
+	var err error
+	if err := os.MkdirAll(filepath.Join(dbPath, chainName), 0755); err != nil {
+		Log.Fatal("mkdir ", dbPath, " failed: ", err)
+	}
+	c.blocksFile, err = os.OpenFile(c.blocksName, os.O_CREATE|os.O_RDWR|os.O_APPEND, 0644)
+	if err != nil {
+		Log.Fatal("open ", c.blocksName, " failed: ", err)
+	}
+	c.lengthsFile, err = os.OpenFile(c.lengthsName, os.O_CREATE|os.O_RDWR|os.O_APPEND, 0644)
+	if err != nil {
+		Log.Fatal("open ", c.lengthsName, " failed: ", err)
+	}
+	if redownload {
+		c.truncZero()
+	}
+	lengths, err := ioutil.ReadFile(c.lengthsName)
+	if err != nil {
+		Log.Fatal("read ", c.lengthsName, " failed: ", err)
+	}
+	if (len(lengths) % 4) > 0 {
+		// some kind of corruption, start over
+		Log.Warningln("lengths file has unexpected length, re-downloading")
+		c.truncZero()
+		lengths = nil
+	}
+
+	// Because the ends of these files are more likely to have corruption
+	// due to (for example) unclean shutdown, let's discard the last several
+	// blocks (re-download them from zcashd).
+	n := int(len(lengths)/4) - backup
+	if n < 0 {
+		n = 0
 	}
+
+	// The last entry in starts[] is where to write the next block.
+	var offset int64
+	c.starts = nil
+	c.starts = append(c.starts, 0)
+	var i int
+	for i = 0; i < n; i++ {
+		length := binary.LittleEndian.Uint32(lengths[i*4 : (i+1)*4])
+		if length < 78 || length > 8*1000*1000 {
+			// start over
+			Log.Warningln("lengths file has impossible value, re-downloading")
+			c.truncZero()
+			offset = 0
+			break
+		}
+		offset += int64(length) + 8
+		c.starts = append(c.starts, offset)
+		if i == 0 {
+			// first block, sanity check it has Sapling activation height
+			block := c.readBlock(c.firstBlock)
+			if block == nil || int(block.Height) != startHeight {
+				// start over
+				Log.Warningln("first block looks wrong, re-downloading")
+				c.truncZero()
+				offset = 0
+				break
+			}
+		}
+		c.nextBlock++
+	}
+
+	// Make sure the files have the correct length (if unclean shutdown)
```

Comment:
Requested change: If the truncation actually would change the file, there should be some kind of error log generated here.

Reply:
An easy way to do that might be to try reading a byte from the blocks file and check if it's at EOF. The length file will always have the correct length unless we take `break` out of the loop, in which case the file will be truncated to zero length anyways.